control/strategy/pool.go

```diff
@@ -347,21 +359,45 @@ func (p *pool) SubscriptionCount() int {
 }
 
 // SelectAP selects an available plugin from the pool
-func (p *pool) SelectAP(taskID string) (SelectablePlugin, serror.SnapError) {
+func (p *pool) SelectAP(opts SelectorValues) (AvailablePlugin, serror.SnapError) {
 	p.RLock()
 	defer p.RUnlock()
 
-	sp := make([]SelectablePlugin, p.Count())
-	i := 0
-	for _, plg := range p.plugins {
-		sp[i] = plg
-		i++
+	selector, err := func(opts SelectorValues) (string, error) {
+		var e error
+		s := ""
+		switch p.Strategy().String() {
+		case "least-recently-used":
+			s = "lru"
+		case "sticky":
+			s = opts.Task
+		case "config-based":
+			s = idFromCfg(opts.Config)
+		default:
+			e = ErrBadStrategy
+		}
+		return s, e
+	}(opts)
+
+	if err != nil {
+		return nil, serror.New(err)
 	}
-	sap, err := p.Select(sp, taskID)
-	if err != nil || sap == nil {
+	aps := p.plugins.Values()
+	ap, err := p.Select(aps, selector)
+	if err != nil || ap == nil {
```

Comment:
If err and ap are nil this will a panic when someone tries to get the error from that serror. 

If both being nil isn't possible, I don't think both checks are needed.